### DISCUSSION TITLE
Can't play a loop without gaps...

### DISCUSSION THREAD

**Discussion by massimag** [WebAudio/web-audio-api#2505] [Q&A]
I need to control the volumes of various audio loops being played, and am trying to do this with webaudio. The web application must work on both Android and IOS:
The files have been edited on a DAW in order to be perfect loops. However, during webaudio playback there is always a gap between the end and the beginning. The gap appears variable and device dependent. If multiple files are played, the gaps often interrupt the whole mix (so even the files that are far from their ends).
At the moment I got the best behavior using AudioBuffer (see minimal code below). Using ogg instead of mp3 the situation improves (although not entirely). However webaudio does not seem to play ogg files on iOS. :-(
I can't find a solution. Is there a webaudio code example that plays loops without any glitches on both Android and iOS?

<!DOCTYPE html>
<html>
<head>
  <script>
  
  var source;
  var audioBuffer;
  var audioCtx
  createAudio = async() =>{
    audioCtx = new AudioContext();
    source = audioCtx.createBufferSource();
    const arrayBuffer = await fetch(
        "test.mp3",
    ).then((res) => res.arrayBuffer());	
    audioBuffer = await audioCtx.decodeAudioData(arrayBuffer);
  }
  
  function startAudio(){
    source.buffer = audioBuffer;
    source.loop = true;
    source.connect(audioCtx.destination);
    source.start();
  }

</script>
 </head>

 <body onClick="startAudio();" onload="createAudio();">    
TEST
<br>
</html>

**Comment by guest271314** [comment]
Have you tried using `AudioWorkletNode`?

**Comment by tgrajewski** [comment]
MP3 files are not generally gapless, they contain some padding samples (there are some workarounds). The AAC files might work better and if possible try Opus, which is truly gapless.

**Reply by guest271314** [reply]
> MP3 files are not generally gapless, th

I write gapless MP3 (and Opus in WebM container) files in the browser here https://github.com/guest271314/captureSystemAudio/blob/master/native_messaging/capture_system_audio/background.js,  and Opus packets produced by WebCodecs `AudioEncoder` without a container here https://github.com/guest271314/WebCodecsOpusRecorder.

**Comment by padenot** [comment]
Those are browser bugs, please report the bugs to the browser vendors. mp3 files can be gapless, aac files can also be gapless, what matters is how the browser uses the codec implementation, and more importantly, if it skips the encoder delay and the padding appropriately.

**Reply by guest271314** [reply]
Are we sure there is a bug? If the original recording contains silence at beginning and/or end that can be perceived as a gap in playback when loop is used.

**Reply by padenot** [reply]
This is not what is being discussed here, the original poster said:

> The files have been edited on a DAW in order to be perfect loops.

**Reply by guest271314** [reply]
> Using ogg instead of mp3 the situation improves (although not entirely).

**Reply by padenot** [reply]
Yes, because OGG has been historically been integrated in browsers in a way that already trims the priming samples.

**Reply by guest271314** [reply]
My point is if there is a perfect loop then how to MP3 and OGG get involved at all? 

> However webaudio does not seem to play ogg files on iOS. :-(

You think OP decodes OGG or MP3 using Web Audio API and however iOS is decoding the the file is the issue? How do we reproduce to verify?

**Reply by padenot** [reply]
Please educate yourself about encoder delay and priming samples for compressed audio formats. Some resources, per codec:

- mp3: https://lame.sourceforge.io/tech-FAQ.txt
- aac: https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/QTFFAppenG/QTFFAppenG.html
- vorbis: https://xiph.org/vorbis/doc/Vorbis_I_spec.html#x1-800004.3.8

Some implementations don't perform the correct trimming at the beginning or the end, this is known, and will be fixed eventually.

**Reply by guest271314** [reply]
So the bug is the encoder, not the browser? Steps to reproduce?

**Reply by guest271314** [reply]
How is it that I can encode to MP3 or Opus in WebM or Opus packets produced by WebCodecs without a container and playback the media in the browser in a loop without gaps? Is the bug specific to Apple devices only?

**Reply by padenot** [reply]
The bug is in the browser (or the decoding side, in general).

I've filed https://github.com/w3c/webcodecs/issues/626 for Web Codecs, it can work in some cases, but there is no general solution to make it work with all codecs.

**Reply by guest271314** [reply]
> If PCM is encoded with a particular codec, and then decoded, does it roundtrip properly? I don't think it can, because the number of priming samples is not returned by the encoder, so there's no way to mux the file properly to signal this (e.g., using an edit list in an mp4, or any other codec/container pair specific way).

Interesting. 

So we don't have to deal with containers I encode WebCodecs `'opus'` packets into a single file without a container and play back the file "properly" using Media Source Extensions and/or converting to WAV for `HTMLMediaElement` playback, or use `AudioWorklet`. 

I keep track of the packet offsets, encode the offsets, WebCodecs metadata , including optional Media Session `MediaMetadata` with images, in the single file. I write the "configuration" header as a `Uint32Array` - the length of the JSON following - then get the subarray using offsets.

I did notice the `8`s written to the beginning of packets from `AudioEncoder` as `Uint8Array` values.  Is that your "priming samples" produced by WebCodecs?

For the case of the output of `decodeAudioData()` can't we just trim all `0`'s from the beginning and end of the channels?

**Reply by guest271314** [reply]
> Please educate yourself about encoder delay and priming samples for compressed audio formats. Some resources, per codec:

> it can work in some cases, but there is no general solution to make it work with all codecs.

WebCodecs was supposed to solve all of those inadequacies. You are in complete control of the encoder and decoder source code, so you can't blame the codec anymore. Or, am I missing something?

**Reply by padenot** [reply]
We're certainly not in control of the codec implementation, some codecs are patented and some implementations have to use the OS library.

More over, the encoder delay can vary, it can depend on the implementation. `ffmpeg` AAC encoder has usually 1024 frames of encoder delay, while Apple encoder has usually 2112 frames of encoder delay.

We need to have Web Codecs output the exact number each time for this reason, we're adding this to Web Codecs.

**Reply by guest271314** [reply]
> We're certainly not in control of the codec implementation, some codecs are patented and some implementations have to use the OS library.

Opus is FOSS. The last time I checked the inventor of MP3 said the code is free to use unemcumbered. FFmpeg is FOSS. WebCodecs can certainly change the code.

> More over, the encoder delay can vary, it can depend on the implementation.

I'm talking about the output of WebCodecs AudioEncoder. Which can be saved to a single file and played back.

**Comment by massimag** [comment]
Thanks for all replies.

In the end my solution was to use very long samples, and do a fade at the beginning and at the end. Obviously the fade is perceived, but not often, and the final experience was acceptable for the user.

Please note this: webaudio does not seem to play ogg files on iOS (or, at least, I had no success).

**Reply by padenot** [reply]
This is expected, iOS (and by extension Safari and other browsers on iOS) don't support Vorbis, which is the codec most often included in OGG containers.
